File: library/src/main/java/com/nextstory/widget/DataBindingGridLayout.java
package com.nextstory.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;
import androidx.databinding.BindingAdapter;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.nextstory.R;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.WeakHashMap;

/**
 * 데이터바인딩 그리드 레이아웃
 *
 * @author troy
 * @version 1.0
 * @since 1.1
 */
@SuppressWarnings("UnusedDeclaration")
public final class DataBindingGridLayout extends RecyclerView {
    private int orientation;

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    private InternalAdapter adapter;

    @BindingAdapter("onBottomScrollReached")
    public static void addOnBottomScrollReachedListener(
            @NonNull final DataBindingGridLayout view,
            @Nullable final OnBottomScrollReachedListener l
    ) {
        view.addOnScrollListener(new OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView view, int newState) {
                if (!view.canScrollVertically(1) && newState == RecyclerView.SCROLL_STATE_IDLE) {
                    if (l != null) {
                        l.onBottomScrollReached();
                    }
                }
            }
        });
    }

    public DataBindingGridLayout(@NonNull Context context) {
        super(context);
        initialize(context, null);
    }

    public DataBindingGridLayout(@NonNull Context context,
                                 @Nullable AttributeSet attrs) {
        super(context, attrs);
        initialize(context, attrs);
    }

    public DataBindingGridLayout(@NonNull Context context,
                                 @Nullable AttributeSet attrs,
                                 int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initialize(context, attrs);
    }

    /**
     * 초기화
     *
     * @param context 컨텍스트
     * @param attrs   AttributeSet
     */
    private void initialize(@NonNull Context context, @Nullable AttributeSet attrs) {
        boolean reverseLayout;
        int spanCount;

        if (attrs == null) {
            orientation = VERTICAL;
            spanCount = 1;
            reverseLayout = false;
        } else {
            TypedArray a =
                    context.obtainStyledAttributes(attrs, R.styleable.DataBindingGridLayout);
            orientation = a.getInt(
                    R.styleable.DataBindingGridLayout_android_orientation,
                    VERTICAL);
            spanCount = a.getInt(
                    R.styleable.DataBindingGridLayout_spanCount,
                    1);
            reverseLayout = a.getBoolean(
                    R.styleable.DataBindingGridLayout_reverseLayout,
                    false);
            a.recycle();
        }

        GridLayoutManager layoutManager = new InternalLayoutManager(
                context, spanCount, orientation, reverseLayout);
        super.setLayoutManager(layoutManager);
        super.setHasFixedSize(false);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (adapter == null) {
            adapter = new InternalAdapter();
            for (int i = 0; i < getChildCount(); i++) {
                View child = getChildAt(i);
                if (child instanceof DataBindingGridLayoutItem) {
                    ((DataBindingGridLayoutItem) child).setParent(this);
                }
                adapter.add(child);
            }
            removeAllViews();
            super.setAdapter(adapter);
        }
    }

    @Override
    protected boolean checkLayoutParams(ViewGroup.LayoutParams p) {
        return p instanceof LayoutParams;
    }

    @Override
    protected LayoutParams generateDefaultLayoutParams() {
        if (orientation == VERTICAL) {
            return new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
        } else {
            return new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.MATCH_PARENT);
        }
    }

    @Override
    public LayoutParams generateLayoutParams(AttributeSet attrs) {
        return new LayoutParams(getContext(), attrs);
    }

    @Override
    protected LayoutParams generateLayoutParams(ViewGroup.LayoutParams p) {
        return new LayoutParams(p);
    }

    @Nullable
    @Override
    public final Adapter<?> getAdapter() {
        return super.getAdapter();
    }

    @Override
    public final void setAdapter(@Nullable Adapter adapter) {
        // no-op
    }

    @Nullable
    @Override
    public final LayoutManager getLayoutManager() {
        return super.getLayoutManager();
    }

    @Override
    public final void setLayoutManager(@Nullable LayoutManager layout) {
        // no-op
    }

    @Override
    public final void setHasFixedSize(boolean hasFixedSize) {
        // no-op
    }

    @Override
    public final boolean hasFixedSize() {
        return false;
    }

    public int getOrientation() {
        return orientation;
    }

    /**
     * 내부 {@link GridLayoutManager}
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    static class InternalLayoutManager extends GridLayoutManager {
        public InternalLayoutManager(Context context,
                                     int spanCount,
                                     int orientation,
                                     boolean reverseLayout) {
            super(context, spanCount, orientation, reverseLayout);
        }

        @Override
        public boolean supportsPredictiveItemAnimations() {
            return false;
        }

        @Override
        public void onLayoutChildren(Recycler recycler, State state) {
            try {
                super.onLayoutChildren(recycler, state);
            } catch (IndexOutOfBoundsException ignore) {
                // no-op
            }
        }
    }

    /**
     * 내부 어댑터
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    static class InternalAdapter extends RecyclerView.Adapter<InternalViewHolder> {
        private final List<View> views = new ArrayList<>();
        private final Map<View, Integer> spanCounts = new WeakHashMap<>();

        @Override
        public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
            GridLayoutManager layoutManager = (GridLayoutManager) recyclerView.getLayoutManager();
            if (layoutManager != null) {
                layoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                    @Override
                    public int getSpanSize(int position) {
                        if (views.size() == 0) {
                            return 1;
                        }
                        View view = views.get(getItemViewType(position));
                        if (view.getLayoutParams() == null) {
                            return 1;
                        }
                        if (spanCounts.containsKey(view)) {
                            return Objects.requireNonNull(spanCounts.get(view));
                        }
                        return 1;
                    }
                });
            }
        }

        @NonNull
        @Override
        public InternalViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = views.get(viewType);
            if (view instanceof DataBindingGridLayoutItem) {
                if (parent.isInEditMode()) {
                    return new InternalViewHolder(view);
                }
                DataBindingGridLayoutItem v = (DataBindingGridLayoutItem) view;
                ViewDataBinding binding = DataBindingUtil.inflate(
                        LayoutInflater.from(parent.getContext()),
                        v.getLayoutRes(),
                        null,
                        false);
                binding.getRoot().setVisibility(View.GONE);
                return new InternalViewHolder(binding, v);
            } else {
                ViewParent viewParent = view.getParent();
                if (viewParent != null) {
                    if (viewParent instanceof ViewGroup) {
                        ViewGroup viewGroup = (ViewGroup) viewParent;
                        viewGroup.removeView(view);
                    }
                }
                return new InternalViewHolder(view);
            }
        }

        @Override
        public void onBindViewHolder(@NonNull InternalViewHolder holder, int position) {
            if (holder.itemView.isInEditMode()) {
                return;
            }
            if (holder.binding == null) {
                return;
            }
            int viewType = getItemViewType(position);
            int realPosition = getRealPosition(position);
            if (!(views.get(viewType) instanceof DataBindingGridLayoutItem)) {
                return;
            }
            DataBindingGridLayoutItem gridItem = (DataBindingGridLayoutItem) views.get(viewType);
            if (gridItem.getItems().size() < realPosition) {
                return;
            }
            holder.binding.getRoot().post(() ->
                    holder.binding.getRoot().setVisibility(View.VISIBLE));
            if (holder.getItemVarId() != -1) {
                holder.binding.setVariable(
                        holder.getItemVarId(),
                        gridItem.getItems().get(realPosition));
            }
            if (holder.getPositionVarId() != -1) {
                holder.binding.setVariable(
                        holder.getPositionVarId(),
                        realPosition);
            }
            if (holder.getCallbackVarId() != -1) {
                holder.binding.setVariable(
                        holder.getCallbackVarId(),
                        (DataBindingGridLayoutItem.Callback) (v, o, i) -> {
                            DataBindingGridLayoutItem.Callback callback = gridItem.getCallback();
                            if (callback != null) {
                                callback.onItemCallback(v, o, i);
                                return;
                            }
                            for (View view : views) {
                                if (view instanceof DataBindingGridLayoutItem) {
                                    DataBindingGridLayoutItem item =
                                            (DataBindingGridLayoutItem) view;
                                    if (item.getCallback() != null) {
                                        item.getCallback().onItemCallback(v, o, i);
                                    }
                                }
                            }
                        });
            }
        }

        @Override
        public int getItemCount() {
            int itemCount = 0;
            for (View view : views) {
                if (view instanceof DataBindingGridLayoutItem) {
                    DataBindingGridLayoutItem gridItem = (DataBindingGridLayoutItem) view;
                    List<?> items = gridItem.getItems();
                    itemCount += items.size();
                } else {
                    itemCount++;
                }
            }
            return itemCount;
        }

        @Override
        public int getItemViewType(int position) {
            int currentPosition = 0;
            for (int type = 0; type < views.size(); type++) {
                View view = views.get(type);
                int itemCount;
                if (view instanceof DataBindingGridLayoutItem) {
                    DataBindingGridLayoutItem gridItem = (DataBindingGridLayoutItem) view;
                    List<?> items = gridItem.getItems();
                    itemCount = items.size();
                } else {
                    itemCount = 1;
                }
                for (int i = 0; i < itemCount; i++) {
                    if (currentPosition == position) {
                        return type;
                    }
                    currentPosition++;
                }
            }
            return 0;
        }

        @Override
        public long getItemId(int position) {
            int realPosition = getRealPosition(position);
            int viewType = getItemViewType(position);
            View view = views.get(viewType);
            if (view instanceof DataBindingGridLayoutItem) {
                DataBindingGridLayoutItem gridItem = (DataBindingGridLayoutItem) view;
                List<?> items = gridItem.getItems();
                return items.get(realPosition).hashCode();
            } else {
                return view.hashCode();
            }
        }

        public int getRealPosition(int position) {
            int currentPosition = 0;
            for (int type = 0; type < views.size(); type++) {
                View view = views.get(type);
                int itemCount;
                if (view instanceof DataBindingGridLayoutItem) {
                    DataBindingGridLayoutItem gridItem = (DataBindingGridLayoutItem) view;
                    List<?> items = gridItem.getItems();
                    itemCount = items.size();
                } else {
                    itemCount = 1;
                }
                for (int i = 0; i < itemCount; i++) {
                    if (currentPosition == position) {
                        return i;
                    }
                    currentPosition++;
                }
            }
            return currentPosition;
        }

        /**
         * 추가
         *
         * @param view 뷰
         */
        private void add(View view) {
            views.add(view);
            spanCounts.put(view, ((LayoutParams) view.getLayoutParams()).getSpanCount());
        }
    }

    /**
     * 내부 뷰 홀더
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    static class InternalViewHolder extends ViewHolder {
        private final ViewDataBinding binding;
        private int itemVarId = -1;
        private int positionVarId = -1;
        private int callbackVarId = -1;

        public InternalViewHolder(@NonNull ViewDataBinding binding,
                                  @NonNull DataBindingGridLayoutItem vh) {
            super(binding.getRoot());
            this.binding = binding;
            String packageName = vh.getContext().getPackageName();
            itemVarId = getBindingId(
                    packageName.concat(".BR.").concat(vh.getItemBindingName()));
            positionVarId = getBindingId(
                    packageName.concat(".BR.").concat(vh.getPositionBindingName()));
            callbackVarId = getBindingId(
                    packageName.concat(".BR.").concat(vh.getCallbackBindingName()));
        }

        public InternalViewHolder(@NonNull View itemView) {
            super(createView(itemView));
            this.binding = null;
        }

        private static int getBindingId(String id) {
            String className = id.substring(0, id.lastIndexOf("."));
            String fieldName = id.substring(id.lastIndexOf(".") + 1);
            try {
                Class<?> brClass = Class.forName(className);
                Field varField = brClass.getField(fieldName);
                return (int) Objects.requireNonNull(varField.get(null));
            } catch (Throwable ignore) {
                return -1;
            }
        }

        private static View createView(View v) {
            Context context = v.getContext();
            ViewGroup.MarginLayoutParams params = (MarginLayoutParams) v.getLayoutParams();
            FrameLayout.LayoutParams newParams =
                    new FrameLayout.LayoutParams(params.width, params.height);
            newParams.topMargin = params.topMargin;
            newParams.bottomMargin = params.bottomMargin;
            newParams.leftMargin = params.leftMargin;
            newParams.rightMargin = params.rightMargin;
            FrameLayout newParent = new FrameLayout(context);
            newParent.addView(v, newParams);
            return newParent;
        }

        public int getItemVarId() {
            return itemVarId;
        }

        public int getPositionVarId() {
            return positionVarId;
        }

        public int getCallbackVarId() {
            return callbackVarId;
        }
    }

    /**
     * 하위 레이아웃 속성 클래스
     */
    public static class LayoutParams extends GridLayoutManager.LayoutParams {
        private int spanCount;

        public LayoutParams(Context c, AttributeSet attrs) {
            super(c, attrs);
            resolveAttributes(c, attrs);
        }

        public LayoutParams(int width, int height) {
            super(width, height);
        }

        public LayoutParams(ViewGroup.LayoutParams source) {
            super(source);
        }

        private void resolveAttributes(@NonNull Context c, @Nullable AttributeSet attrs) {
            if (attrs == null) {
                spanCount = 1;
            } else {
                TypedArray a = c.obtainStyledAttributes(
                        attrs, R.styleable.DataBindingGridLayout_Layout);
                spanCount = a.getInt(
                        R.styleable.DataBindingGridLayout_Layout_spanCount,
                        1);
                a.recycle();
            }
        }

        public int getSpanCount() {
            return spanCount;
        }
    }

    /**
     * 하단 스크롤 리스너
     */
    public interface OnBottomScrollReachedListener {
        /**
         * 하단 스크롤 도달 시 호출
         */
        void onBottomScrollReached();
    }
}

File: library/src/main/java/com/nextstory/widget/DataBindingGridLayout.kt
@file:Suppress(
  "unused",
  "SpellCheckingInspection",
  "MemberVisibilityCanBePrivate",
  "RedundantModalityModifier",
  "NotifyDataSetChanged",
  "RedundantGetter",
)

package com.nextstory.widget

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.annotation.RestrictTo
import androidx.databinding.BindingAdapter
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.nextstory.BR
import com.nextstory.R
import java.lang.ref.WeakReference

/**
 * 데이터바인딩을 지원하는 그리드 형식의 레이아웃
 *
 * 'app:spanCount' 속성으로 레이아웃의 열, 행 개수를 지정합니다. 하위 뷰에서는
 * 'app:dbv_layout_span' 속성으로 차지하는 개수를 지정할 수 있습니다.
 *
 * 예시:
 * ```xml
 * <com.nextstory.widget.DataBindingGridLayout
 *   android:layout_width="match_parent"
 *   android:layout_height="match_parent"
 *   app:dbv_spanCount="1">
 *
 *   <TextView
 *     android:layout_width="wrap_content"
 *     android:layout_height="wrap_content"
 *     android:text="hello"
 *     app:dbv_layout_span="1" />
 *
 *   <com.nextstory.widget.DataBindingGridLayoutItem
 *     android:layout_width="wrap_content"
 *     android:layout_height="wrap_content"
 *     app:dbv_layout_span="1"
 *     app:dbv_items="@{activity.testItems}"
 *     app:dbv_layout="@layout/item_layout_id"
 *     app:dbv_onItemCallback="@{(v, item, position) -> activity.onCallback(position)}"/>
 *
 * </com.nextstory.widget.DataBindingGridLayout>
 * ```
 */
class DataBindingGridLayout : RecyclerView {
  var orientation: Int = VERTICAL
    get() = field
    set(value) {
      field = value

      if (layoutManager != null && layoutManager is InternalLayoutManager) {
        (layoutManager as InternalLayoutManager).orientation = field
        adapter?.notifyDataSetChanged()
      }
    }

  var spanCount: Int = 1
    get() = field
    set(value) {
      field = value

      if (layoutManager != null && layoutManager is InternalLayoutManager) {
        (layoutManager as InternalLayoutManager).spanCount = field
        adapter?.notifyDataSetChanged()
      }
    }

  var reverseLayout: Boolean = false
    get() = field
    set(value) {
      field = value

      if (layoutManager != null && layoutManager is InternalLayoutManager) {
        (layoutManager as InternalLayoutManager).reverseLayout = field
        adapter?.notifyDataSetChanged()
      }
    }

  constructor(context: Context) : super(context) {
    init(context, null)
  }

  constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
    init(context, attrs)
  }

  constructor(context: Context, attrs: AttributeSet?, res: Int) : super(context, attrs, res) {
    init(context, attrs)
  }

  private fun init(context: Context, attrs: AttributeSet?) {
    if (attrs == null) {
      orientation = VERTICAL
      spanCount = 1
      reverseLayout = false
    } else {
      val a = context.obtainStyledAttributes(attrs, R.styleable.DataBindingGridLayout)
      orientation = a.getInt(R.styleable.DataBindingGridLayout_dbv_orientation, VERTICAL)
      spanCount = a.getInt(R.styleable.DataBindingGridLayout_dbv_spanCount, 1)
      reverseLayout = a.getBoolean(R.styleable.DataBindingGridLayout_dbv_reverseLayout, false)
      a.recycle()
    }

    super.setLayoutManager(
      InternalLayoutManager(
        context,
        spanCount,
        orientation,
        reverseLayout,
      )
    )

    super.setItemAnimator(null)
    super.setHasFixedSize(false)
  }

  override fun checkLayoutParams(p: ViewGroup.LayoutParams?): Boolean {
    return p is LayoutParams
  }

  override fun generateDefaultLayoutParams(): ViewGroup.LayoutParams {
    return if (orientation == VERTICAL) LayoutParams(-1, -2) else LayoutParams(-2, -1)
  }

  override fun generateLayoutParams(attrs: AttributeSet?): ViewGroup.LayoutParams {
    return LayoutParams(context, attrs)
  }

  override fun generateLayoutParams(p: ViewGroup.LayoutParams?): ViewGroup.LayoutParams {
    return LayoutParams(p)
  }

  final override fun getAdapter(): Adapter<*>? {
    return super.getAdapter()
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY)
  final override fun setAdapter(adapter: Adapter<*>?) {
    // NOOP: 내부 구현
  }

  final override fun getLayoutManager(): LayoutManager? {
    return super.getLayoutManager()
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY)
  final override fun setLayoutManager(layout: LayoutManager?) {
    // NOOP: 내부 구현
  }

  final override fun hasFixedSize(): Boolean {
    return false
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY)
  final override fun setHasFixedSize(hasFixedSize: Boolean) {
    // NOOP: 내부 구현
  }

  @CallSuper
  override fun onAttachedToWindow() {
    super.onAttachedToWindow()

    if (adapter == null) {
      val newAdapter = InternalDataBindingGridLayoutAdapter()

      for (i in 0 until childCount) {
        val childView = getChildAt(i)
        if (childView is DataBindingViewItem) {
          childView.owner = WeakReference(this)
        }
        newAdapter.views.add(childView)
      }

      removeAllViews()
      super.setAdapter(newAdapter)
    }
  }

  @Suppress("OVERRIDE_DEPRECATION", "DEPRECATION")
  class LayoutParams : GridLayoutManager.LayoutParams {
    var span: Int = 1

    constructor(width: Int, height: Int) : super(width, height)
    constructor(source: ViewGroup.LayoutParams?) : super(source)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
      if (attrs != null) {
        val a = context.obtainStyledAttributes(attrs, R.styleable.DataBindingGridLayout_Layout)
        span = a.getInt(R.styleable.DataBindingGridLayout_Layout_dbv_layout_span, 1)
        a.recycle()
      }
    }

    override fun viewNeedsUpdate(): Boolean {
      return try {
        super.viewNeedsUpdate()
      } catch (ignored: Exception) {
        false
      }
    }

    override fun isViewInvalid(): Boolean {
      return try {
        super.isViewInvalid()
      } catch (ignored: Exception) {
        false
      }
    }

    override fun isItemRemoved(): Boolean {
      return try {
        super.isItemRemoved()
      } catch (ignored: Exception) {
        false
      }
    }

    override fun isItemChanged(): Boolean {
      return try {
        super.isItemRemoved()
      } catch (ignored: Exception) {
        false
      }
    }

    override fun getViewPosition(): Int {
      return try {
        super.getViewPosition()
      } catch (ignored: Exception) {
        0
      }
    }

    override fun getViewLayoutPosition(): Int {
      return try {
        super.getViewLayoutPosition()
      } catch (ignored: Exception) {
        0
      }
    }

    override fun getViewAdapterPosition(): Int {
      return try {
        super.getViewAdapterPosition()
      } catch (ignored: Exception) {
        0
      }
    }

    override fun getSpanSize(): Int {
      return span
    }
  }
}

@RestrictTo(RestrictTo.Scope.LIBRARY)
class InternalLayoutManager(
  context: Context?,
  spanCount: Int,
  orientation: Int,
  reverseLayout: Boolean,
) : GridLayoutManager(context, spanCount, orientation, reverseLayout) {
  override fun supportsPredictiveItemAnimations(): Boolean {
    return false
  }

  override fun onLayoutChildren(recycler: RecyclerView.Recycler?, state: RecyclerView.State?) {
    try {
      super.onLayoutChildren(recycler, state)
    } catch (_: IndexOutOfBoundsException) {
      // NOOP: ignored
    }
  }
}

@RestrictTo(RestrictTo.Scope.LIBRARY)
class InternalDataBindingGridLayoutAdapter
  : RecyclerView.Adapter<InternalDataBindingGridLayoutAdapter.InternalViewHolder>() {
  val views = mutableListOf<View>()

  init {
    setHasStableIds(true)
  }

  override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
    if (recyclerView.layoutManager is InternalLayoutManager) {
      val manager = recyclerView.layoutManager as InternalLayoutManager
      manager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
        override fun getSpanSize(position: Int): Int {
          if (views.size == 0) {
            return 1
          }

          val view = views[getItemViewType(position)]
          if (view.layoutParams == null
            || view.layoutParams !is DataBindingGridLayout.LayoutParams
          ) {
            return 1
          }

          val params = view.layoutParams as DataBindingGridLayout.LayoutParams
          return params.span
        }
      }
    }
  }

  override fun getItemCount(): Int {
    var itemCount = 0
    for (view in views) {
      if (view is DataBindingViewItem && view.items != null) {
        itemCount += view.items!!.size
      } else {
        itemCount++
      }
    }
    return itemCount
  }

  override fun getItemViewType(position: Int): Int {
    var currentPosition = 0
    for (type in views.indices) {
      val view = views[type]
      val itemCount = if (view is DataBindingViewItem && view.items != null) {
        view.items!!.size
      } else {
        1
      }

      for (i in 0 until itemCount) {
        if (currentPosition == position) {
          return type
        }
        currentPosition++
      }
    }
    return 0
  }

  override fun getItemId(position: Int): Long {
    val realPosition = getRealPosition(position)
    val viewType = getItemViewType(position)
    val view = views[viewType]
    return if (view is DataBindingViewItem && view.items != null) {
      view.items!![realPosition].hashCode().toLong()
    } else {
      view.hashCode().toLong()
    }
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): InternalViewHolder {
    val view = views[viewType]

    if (view is DataBindingViewItem) {
      val params = view.layoutParams as DataBindingGridLayout.LayoutParams
      params.leftMargin = 0
      params.topMargin = 0
      params.rightMargin = 0
      params.bottomMargin = 0

      if (view.isInEditMode) {
        return InternalViewHolder(view)
      }

      val b: ViewDataBinding = DataBindingUtil.inflate(
        LayoutInflater.from(parent.context),
        view.layoutRes,
        null,
        false,
      )

      // NOTE: 데이터바인딩을 이용하는 뷰는 비정상적으로 보여지는 현상이 있어 잠시 숨깁니다.
      b.root.visibility = RecyclerView.GONE

      return InternalViewHolder(b)
    }

    val viewParent = view.parent
    if (viewParent != null) {
      if (viewParent is ViewGroup) {
        viewParent.removeView(view)
      }
    }

    return InternalViewHolder(view)
  }

  override fun onBindViewHolder(holder: InternalViewHolder, position: Int) {
    if (holder.itemView.isInEditMode) {
      return
    }

    if (holder.binding == null) {
      return
    }

    val viewType = getItemViewType(position)
    if (views[viewType] !is DataBindingViewItem) {
      return
    }

    val realPosition = getRealPosition(position)
    val gridItem = views[viewType] as DataBindingViewItem
    if (gridItem.items == null || gridItem.items!!.size < realPosition) {
      return
    }

    // NOTE: 임시로 숨겨졌던 뷰를 다시 표시합니다.
    holder.binding.root.post { holder.binding.root.visibility = RecyclerView.VISIBLE }

    holder.binding.setVariable(BR.item, gridItem.items!![realPosition])
    holder.binding.setVariable(BR.position, realPosition)
    holder.binding.setVariable(BR.callback, object : DataBindingItemCallback {
      override fun onDataBindingItemCallback(view: View, item: Any?, position: Int) {
        if (gridItem.callback != null) {
          gridItem.callback!!.onDataBindingItemCallback(view, item, position)
          return
        }

        for (v in views) {
          if (v is DataBindingViewItem) {
            if (v.callback != null) {
              v.callback!!.onDataBindingItemCallback(view, item, position)
            }
          }
        }
      }
    })
  }

  private fun getRealPosition(position: Int): Int {
    var currentPosition = 0
    for (type in views.indices) {
      val view = views[type]
      val itemCount = if (view is DataBindingViewItem && view.items != null) {
        view.items!!.size
      } else {
        1
      }

      for (i in 0 until itemCount) {
        if (currentPosition == position) {
          return i
        }
        currentPosition++
      }
    }
    return currentPosition
  }

  class InternalViewHolder : RecyclerView.ViewHolder {
    val binding: ViewDataBinding?

    constructor(itemView: View) : super(wrapView(itemView)) {
      this.binding = null
    }

    constructor(binding: ViewDataBinding) : super(binding.root) {
      this.binding = binding
    }

    companion object {
      fun wrapView(view: View): View {
        return view
      }
    }
  }
}

object DataBindingGridLayoutBindingAdapter {
  @JvmStatic
  @BindingAdapter("dbv_spanCount")
  fun setSpanCount(view: DataBindingGridLayout, count: Int) {
    if (view.layoutManager != null && view.layoutManager is InternalLayoutManager && count > 0) {
      (view.layoutManager as InternalLayoutManager).spanCount = count
    }
  }

  @JvmStatic
  @BindingAdapter("dbv_layout_span")
  fun setSpanCount(view: View, count: Int) {
    if (view.layoutParams != null && view.layoutParams is DataBindingGridLayout.LayoutParams) {
      (view.layoutParams as DataBindingGridLayout.LayoutParams).span = count
      view.requestLayout()
    }
  }
}
